Proposal for updating IfcPatch ConvertLengthUnit.py for compatibility with boost>=1.82

Krande opened the pull request:
Based on the reported issue in #4116, here is a suggested change for the IfcPatch `ConvertLengthUnit.py` that ensures the intended behaviour is maintained despite changes in the boost>=1.82.

I also moved the core functionality from the patch itself to ifcopenshell-python -> `ifcopenshell.util.unit.convert_file_units(ifc_file, target_unit) -> ifcopenshell.file` so that users of the python library could also enjoy this functionality :)

Let me know what you think?

- closes: #4116

aothms commented on src/ifcopenshell-python/ifcopenshell/util/unit.py:
```diff
@@ -624,3 +622,83 @@ def format_length(
+    return None
+
+
+def convert_file_units(ifc_file: ifcopenshell.file, target_units: str) -> ifcopenshell.file:
```
This probably applies to other APIs as well, but should we copy over the header values?

Krande replied:
Ah, that never occured to me. Do we have any code examples for copying the header values somewhere else?

Krande replied:
In https://github.com/IfcOpenShell/IfcOpenShell/pull/4125/commits/9335c8a0c8c9cbc2fc9056cec79dd0bdd10c7350 I changed the "copy" method to a more direct copy of the entire file instead of going element by element.

`file_patched = ifcopenshell.file.from_string(ifc_file.wrapped_data.to_string())`

Which I believe will include header information

aothms commented on src/ifcopenshell-python/ifcopenshell/util/unit.py:
```diff
@@ -624,3 +622,83 @@ def format_length(
+    prefix = "MILLI" if target_units == "MILLIMETERS" else None
+
+    file_patched = ifcopenshell.api.run("project.create_file", version=ifc_file.schema)
+    if ifc_file.schema == "IFC2X3":
+        user = file_patched.add(ifc_file.by_type("IfcProject")[0].OwnerHistory.OwningUser)
+        application = file_patched.add(ifc_file.by_type("IfcProject")[0].OwnerHistory.OwningApplication)
+        old_get_user = ifcopenshell.api.owner.settings.get_user
+        old_get_application = ifcopenshell.api.owner.settings.get_application
+        ifcopenshell.api.owner.settings.get_user = lambda ifc: user
+        ifcopenshell.api.owner.settings.get_application = lambda ifc: application
```
Is this still needed?

Krande replied:
I just naively copied the IFC2x3 stuff from the patch not really knowing its intended purpose. 

I am fine with removing it

Krande replied:
I just removed it in https://github.com/IfcOpenShell/IfcOpenShell/pull/4125/commits/f5500da0af1ad79a3350bd304bf7b03365b065ea

If anyone wants the IFC2x3 functionality we can add it to the patch instead. Or suggest using the conversion utility to convert from 2x3 to a more recent version :)

aothms commented on src/ifcopenshell-python/ifcopenshell/util/unit.py:
```diff
@@ -624,3 +622,83 @@ def format_length(
+
+    unit_assignment = ifcopenshell.util.unit.get_unit_assignment(file_patched)
+
+    old_length = [u for u in unit_assignment.Units if getattr(u, "UnitType", None) == "LENGTHUNIT"][0]
+    new_length = ifcopenshell.api.run("unit.add_si_unit", file_patched, unit_type="LENGTHUNIT", prefix=prefix)
```
It probably would be a small effort to generalize this beyond length unit, we can save that for later when there is a demand, but maybe make it more explicit in the name of the function.

aothms replied:
Also we can only convert to SI, fine limitation as is, but maybe document.

Krande replied:
Yeah, I think so too. For now I updated the name to make its current purpose more clear in https://github.com/IfcOpenShell/IfcOpenShell/pull/4125/commits/c9552fac29034dae1bec6faa9dcebe446d6b5a5c.

aothms commented on src/ifcopenshell-python/ifcopenshell/util/unit.py:
```diff
@@ -624,3 +622,83 @@ def format_length(
+                # set the new value
+                setattr(element, attr.name(), new_el)
+
+    file_patched.remove(old_length)
```
Does it make more sense to detect this in the loop at line 666 (𝅘𝅥𝅮  number of the beast 𝅘𝅥𝅮) and simply don't add it in the first place?

Krande replied:
Well, we could (for the musical reference alone). However, I kind of liked the simplicity of the "move all elements to new ifc, and do an in-place edit of the new file". 

But I don't really have a strong opinion on this, so please feel free to decide what you think is best :)

Krande commented:
@aothms We should probably include a test for the units conversion. I don't have time to add one today, but I might have some time to add one tomorrow :)

Krande commented:
@aothms I fixed the failing ci.yml and removed ci-py-only-yml (as it was identical to ci.yml just without the IfcConvert tests).

Now all we need to fix the failing tests

aothms commented:
Great 🎉 yeah that ci-py-only was there because at some point IfOpSh didn't build if IfcConvert wasn't configured to be built in CMake, but that probably was a one of event that didn't need a full ci script to run till eternity.
